### DISCUSSION TITLE
Collapsed Records in offline scenario

### DISCUSSION THREAD

**Discussion by henda79** [Azure/azure-mobile-apps#545] [General]
@adrianhall , I recall a conversation we had sometime ago and you explained Collapsed Records ?

On the server side, I'm using tables with FK's to ensure integrity, so as you know, I have to send the tables in order which up to now had worked. However the collapsed records feature make a mess of the order and fails with an exception on the server due to the violated FK.

Is there a way to disable the Collapsed Records feature ?

**Comment by adrianhall** [comment]
There is no way to disable collapsing records.  

One of the things you need to be aware of is that updates are written as they arrive.   If you have two writes to two different tables, then they are written individually without the benefit of transactions.  As a result, establishing FK's to enforce integrity checks will inevitably fail at some point.  It gets worse when you enable parallel writes, but it can happen at any time.  

If you are relying on UpdateA happening before UpdateB, then collapsing records don't change the ordering EXCEPT if parallel writes are turned on.  In that case, UpdateA and UpdateB can be sent to the server in two different threads.  If UpdateB is "smaller" than UpdateA, it will be transmitted quicker, and will likely run first.

**Reply by henda79** [reply]
I'm not really sure how to proceed. Without the FK's on the server I can guarantee a ton of orphaned records.

But if there is no way to guarantee order (I only use a single thread ATM), then I see no other option but to delete the FK's and have a free for all.

It all seemed to be working but then noticed if I updated a table then did not sync and it then got updated later, it would collapse into one and be in the wrong order in the __operations table.

?!?!

**Reply by adrianhall** [reply]
Let's say you have:

* Update-to-tableA-recordX
* Update-to-tableB-recordY

With one thread, these will be sent in order.  Now, let's update them

* Update-to-tableA-recordX-v2
* Update-to-tableB-recordY

The update to tableA will still happen first.  Now, update tableB:

* Update-to-tableA-recordX-v2
* Update-to-tableB-recordY-v2

Again - the order doesn't change - collapsing preserves order in the __operations table.  There is a corner case - when you synchronize while also updating the records.  There is a table lock during the synchronization, so you can't update tableA when a record from tableA is  being sent out of the operations queue, or when tableA is being pulled.  However, that still doesn't change the order.

**Reply by henda79** [reply]
Weird it got out of order for me, heres what I did.

Updated Table Z.

At this point it was the only operation waiting to sync.

I did not sync here.

I inserted rows into tables A, B , C and D
I updated the SAME row in TABLE Z but a different column.

so at this point the operations table is usually in order and shows A, B, C, D, Z

However when I checked the operation table, the 2 updates for table Z where first in the Q (rolled into one, I assume this is the collapsed record part)

So for me the ordering did not work. For it to work, table Z needed to be last in the order.

**Reply by adrianhall** [reply]
So, what you saw was the collapsing working correctly. The two updates to table Z are collapsed with the final order being Z-A-B-C-D.  The two records in Z for the same ID are collapsed into a single record in the original position.

It sounds like your tables are too intertwined for effective offline use and you may need to be selective about what you synchronize - full offline isn't for every situation.  You may, for example, do the updates online and then an immediate pull to get all the changes (i.e. using the offline store as a caching layer).

**Reply by henda79** [reply]
Right, that's what I thought. I was hoping the 2nd update to Table Z would be last in the order where it was performed, then I remembered collapsed records.

All of what we do is in the field and the operator must be able to perform their duty even if there is no signal. We did use Microsoft's SQL CE Merge Replication which worked perfectly but since the technology it worked on is ancient I need to implement a modern solution for android. Ideally I don't want to remove the FK's but it looks like I may have to.

**Reply by amertlich** [reply]
I'm in the process of updating to the latest SDK so I speak only from past experience, but we've been able to keep foreign keys enforced server side (and event client side), but it has introduced some complexity.  

Basically, we handle a collapsed record rejection (because of a foreign key constraint) from the server as a "conflict" that needs to be handled.  We typically handle this in the process of our push logic by removing the foreign key (which we can automate because we know our schema) that caused the conflict (i.e. this would put the record back into the "first" state it was in before the update that added the foreign key violation)  and then pushing the record back to the server. We then make sure "updates" are committed to the operation queue that now include the foreign key (but really the locally persisted record is not changed, it was only pushed with the key removed). We let the remaining pushes complete, and then, as the operation queue is processed, the conflict resolution updates get pushed without failures.  Obviously there could be problems with this depending on your schema, because it doesn't guarantee the same operation order, but this works for our situation.

The only way I could envision handling this in the SDK is if the operations table actually serialized a snapshot version of the item/record so the original version could be pushed. But I would imagine the reason this wasn't done was to avoid bloating the size of the operation table and duplicating the data in the database while data has not yet been synchronized.  Subsequent item updates that are already in the operation queue would only need to store the diff of what's been changed, but at this point you may as well make it a CRDT history/journal. So the way it's implemented does make it simpler for the intended use case, I'd imagine.

**Reply by henda79** [reply]
Thanks for your input. Unfortunately, I'm only one dev tasked with the whole project and my knowledge in not 100% in all areas. Your solution is similar to workarounds I was thinking but I think it would be complex to implement.

How do you enable/disable FK's from the client side ?
When we have a "sync conflict" due to an FK we get a Server Error 500 on the client and no information about it. I have very limited knowledge when it comes to ASP dev.

**Reply by adrianhall** [reply]
A server error 500 is generic "An exception was thrown that we don't know about".  Obviously, I'd need the full exception in order to diagnose the issue.

**Reply by henda79** [reply]
Sorry, I was not clear. I know the error is due to the FK violation on the server side as I can see the logs on the console, but the client just gets back a 500, with no real meaning. Is there anyway to enhance this for the client.

**Reply by adrianhall** [reply]
Write an exception filter for the database errors.  You can find an example here: https://github.com/Azure/azure-mobile-apps/blob/main/sdk/dotnet/src/Microsoft.AspNetCore.Datasync/Filters/DatasyncExceptionsAttribute.cs  In your filter, you can adjust the response any way you want to.  Make sure you are scoping the exception filter to only handle the errors you need to.  (For example, the Datasync service has a HttpException as a core exception that it captures and it ignores all other exceptions).  For example, you can produce a JSON payload for the message and associated information.  You should NOT use a HTTP response code that the client SDK uses (400 would be a good one to use, since that's just Bad Request and is generally passed on to the client application).

Then apply your exception filter as an attribute to your controllers to install it.

**Reply by henda79** [reply]
Thanks Adrian, I'll take a peek.
